mesh/mesh_connections.hpp

```diff
@@ -0,0 +1,276 @@
+// Copyright (c) 2010-2021, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#ifndef MFEM_MESH_CONNECTIONS
+#define MFEM_MESH_CONNECTIONS
+
+#include "../general/array.hpp"
+#include <vector>
+
+namespace mfem
+{
+
+class Table;
+class Mesh;
+
+/** @brief Class that exposes connections between Mesh entities such as
+    vertices, edges, faces, and volumes.  This is done without assuming 
+    the elements have a particular dimension or type and will support 
+    dimensions > 3 in the future.  EntityIndex numbers are described in the 
+    \ref EntityIndex and \ref EntityIndices sturcts.  The connections are defined 
+    on the Mesh object given at the time of construction, and Table objects are built 
+    lazily as they are required so the first query of a paticular kind may be a lot slower
+    than the following queries.  It is worth noting that the our naming convention of child
+    and parent connections actually describes ancestor/decendent relationship eg. a node can 
+    e a "child" of edges, faces, and volumes.**/
+class MeshConnections
+{
+   friend class Mesh;
+
+   private:
+   /// The mesh object that the connections are defined on
+   const Mesh &mesh;
+
+   /** @brief 2D Array of tables describing connections between entity indices in the Mesh.
+       The 2D array represents tables mapping from row indices to column indices of the 3-4
+       different kinds of entities (0D, 1D, 2D, 3D, 2D Boundary) respectively.  For example 
+       the Table mapping Vertices to Boundary Elements/Faces on a 3D mesh would be at T[0][3]*/
+   std::vector<std::vector<Table*>> T;
+
+
+
+   public:
+   /** @brief Struct representing the index number of a geometrical mesh entity.  In 
+       MFEM meshes there is currently support for 5 entity kinds that all have their
+       own index numberings:  vertices, edges, faces, elements, and boundary elements. 
```

Comment:
Should we replace "boundary elements" with "boundary faces"?

Reply:
I'm trying to avoid using the term faces to represent both the 2D things in a 3D mesh and 1D things in a 2D mesh.  Although I am using elements in 2 different ways here now.  I'll see if I an reword it in a way to avoid that.